# IS 362 Week 5 Assignment
## Airports & Weather

Questions to answer:
1. What is the northernmost airport in the United States?
2. What is the easternmost airport in the United States?
3. On February 12th, 2013, which New York area airport had the windiest weather? 

In [1]:
# import modules
import numpy as np
import pandas as pd

Airport geographic is measured in latitude and longitude in the airports csv file. That file will be imported first.

In [2]:
# inputting airport csv file
airports = pd.read_csv('csv_files/airports.csv')
airports.head()

faa                           name        lat        lon   alt  tz dst  \
0  04G              Lansdowne Airport  41.130472 -80.619583  1044  -5   A   
1  06A  Moton Field Municipal Airport  32.460572 -85.680028   264  -6   A   
2  06C            Schaumburg Regional  41.989341 -88.101243   801  -6   A   
3  06N                Randall Airport  41.431912 -74.391561   523  -5   A   
4  09J          Jekyll Island Airport  31.074472 -81.427778    11  -5   A   

              tzone  
0  America/New_York  
1   America/Chicago  
2   America/Chicago  
3  America/New_York  
4  America/New_York

## Question 1
We're looking for the northernmost airport in the United States, so let's first review the time zones included in the csv file:

In [3]:
airports.groupby(['tzone']).count()

faa  name  lat  lon  alt   tz  dst
tzone                                                  
America/Anchorage    239   239  239  239  239  239  239
America/Chicago      342   342  342  342  342  342  342
America/Denver       119   119  119  119  119  119  119
America/Los_Angeles  176   176  176  176  176  176  176
America/New_York     519   519  519  519  519  519  519
America/Phoenix       38    38   38   38   38   38   38
America/Vancouver      2     2    2    2    2    2    2
Asia/Chongqing         2     2    2    2    2    2    2
Pacific/Honolulu      18    18   18   18   18   18   18

Note that there are airports included in the file that are located in Asia, so we'll want to remove that before moving further

In [4]:
airports2 = airports.query('tzone != "Asia/Chongqing"')
airports2.head()

faa                           name        lat        lon   alt  tz dst  \
0  04G              Lansdowne Airport  41.130472 -80.619583  1044  -5   A   
1  06A  Moton Field Municipal Airport  32.460572 -85.680028   264  -6   A   
2  06C            Schaumburg Regional  41.989341 -88.101243   801  -6   A   
3  06N                Randall Airport  41.431912 -74.391561   523  -5   A   
4  09J          Jekyll Island Airport  31.074472 -81.427778    11  -5   A   

              tzone  
0  America/New_York  
1   America/Chicago  
2   America/Chicago  
3  America/New_York  
4  America/New_York

Now we can sort by latitude:

In [5]:
airports2.sort_values(by=['lat'], ascending=False)

faa                                        name        lat         lon  \
417   EEN                     Dillant Hopkins Airport  72.270833   42.898333   
230   BRW                  Wiley Post Will Rogers Mem  71.285446 -156.766003   
110   AIN                          Wainwright Airport  70.638056 -159.994722   
708   K03                               Wainwright As  70.613378 -159.860350   
152   ATK  Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000   
...   ...                                         ...        ...         ...   
930   MUE                               Waimea Kohala  20.001328 -155.668108   
1403  WKL                           Waikoloa Heliport  19.913600 -155.864000   
231   BSF                                Bradshaw Aaf  19.760056 -155.553717   
734   KOA                        Kona Intl At Keahole  19.738767 -156.045631   
679   ITO                                   Hilo Intl  19.721375 -155.048469   

       alt  tz dst              tzone  
417    149  -5   A                NaN  
230     44  -9   A  America/Anchorage  
110     41  -9   A  America/Anchorage  
708     35  -9   A  America/Anchorage  
152     96  -9   A  America/Anchorage  
...    ...  ..  ..                ...  
930   2671 -10   A   Pacific/Honolulu  
1403   109 -10   N   Pacific/Honolulu  
231   6190 -10   A   Pacific/Honolulu  
734     47 -10   N   Pacific/Honolulu  
679     38 -10   N   Pacific/Honolulu  

[1456 rows x 8 columns]

It looks like the first item (Dillant Hopkins Airport) has its latitude and longitude reversed points reversed (it's located in New Hampshire). The northernmost airport is the **Wiley Post–Will Rogers Memorial Airport** with a latitude of 71.285446.

## Question 2
We're looking for the easternmost airport in the United States.
We'll be looking for a longitude with the highest number:

In [6]:
#reusing previously created object
airports2.sort_values(by=['lon'], ascending=False)

faa                        name        lat         lon  alt  tz dst  \
1290  SYA                Eareckson As  52.712275  174.113620   98  -9   A   
417   EEN     Dillant Hopkins Airport  72.270833   42.898333  149  -5   A   
444   EPM  Eastport Municipal Airport  44.910111  -67.012694   45  -5   A   
624   HUL                Houlton Intl  46.123083  -67.792056  489  -5   A   
259   CAR                Caribou Muni  46.871500  -68.017917  626  -5   A   
...   ...                         ...        ...         ...  ...  ..  ..   
1251  SNP              St Paul Island  57.167333 -170.220444   63  -9   A   
1281  SVA            Savoonga Airport  63.686400 -170.493000   53  -9   A   
522   GAM             Gambell Airport  63.766800 -171.733000   27  -9   A   
112   AKB                Atka Airport  52.220278 -174.206389   56  -9   A   
95    ADK                Adak Airport  51.878000 -176.646000   18  -9   A   

                  tzone  
1290  America/Anchorage  
417                 NaN  
444    America/New_York  
624    America/New_York  
259    America/New_York  
...                 ...  
1251  America/Anchorage  
1281  America/Anchorage  
522   America/Anchorage  
112   America/Anchorage  
95    America/Anchorage  

[1456 rows x 8 columns]

Looking at the data, Eareckson Air Station is actually the westernmost airport as it belongs to a different time zone. Also, because the Dillant information is backwards, we will ignore that as well. So the easternmost airport is the **Eastport Municipal Airport** with a longitude of -67.012694.

## Question 3
On February 12th, 2013, which New York area airport had the windiest weather?
For this question, we'll need to import the weather csv file:

In [7]:
# inputting weather csv file
weather = pd.read_csv('csv_files/weather.csv')
weather.head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     1  39.02  26.06  59.37     270.0    10.35702   
1    EWR  2013      1    1     2  39.02  26.96  61.63     250.0     8.05546   
2    EWR  2013      1    1     3  39.02  28.04  64.43     240.0    11.50780   
3    EWR  2013      1    1     4  39.92  28.04  62.21     250.0    12.65858   
4    EWR  2013      1    1     5  39.02  28.04  64.43     260.0    12.65858   

   wind_gust  precip  pressure  visib             time_hour  
0        NaN     0.0    1012.0   10.0  2013-01-01T06:00:00Z  
1        NaN     0.0    1012.3   10.0  2013-01-01T07:00:00Z  
2        NaN     0.0    1012.5   10.0  2013-01-01T08:00:00Z  
3        NaN     0.0    1012.2   10.0  2013-01-01T09:00:00Z  
4        NaN     0.0    1011.9   10.0  2013-01-01T10:00:00Z

To make sure we're looking at the right date, we'll want to filter for the year, month and date:

In [8]:
weather2 = weather[(weather.year == 2013) & (weather.month == 2) & (weather.day == 12)].sort_values(by=['wind_speed'], ascending=False)
weather2

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
1009     EWR  2013      2   12     3  39.02  26.96  61.63     260.0   
18417    LGA  2013      2   12     2  42.98  26.06  50.94     290.0   
1018     EWR  2013      2   12    12  44.06  26.06  48.87     270.0   
18428    LGA  2013      2   12    13  44.06  23.00  43.02     300.0   
18429    LGA  2013      2   12    14  44.06  23.00  43.02     300.0   
...      ...   ...    ...  ...   ...    ...    ...    ...       ...   
1006     EWR  2013      2   12     0  39.92  39.02  96.55     240.0   
1027     EWR  2013      2   12    21  35.96  24.08  61.67     300.0   
18415    LGA  2013      2   12     0  42.08  37.40  86.89       NaN   
9709     JFK  2013      2   12     0  42.08  37.04  82.17     280.0   
1029     EWR  2013      2   12    23  35.96  23.00  58.94     250.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
1009   1048.36058        NaN     0.0    1008.3   10.0  2013-02-12T08:00:00Z  
18417    23.01560   31.07106     0.0    1007.1   10.0  2013-02-12T07:00:00Z  
1018     21.86482   31.07106     0.0    1012.5   10.0  2013-02-12T17:00:00Z  
18428    21.86482   25.31716     0.0    1011.7   10.0  2013-02-12T18:00:00Z  
18429    20.71404   25.31716     0.0    1011.5   10.0  2013-02-12T19:00:00Z  
...           ...        ...     ...       ...    ...                   ...  
1006      6.90468        NaN     0.0    1006.9   10.0  2013-02-12T05:00:00Z  
1027      5.75390        NaN     0.0    1013.0   10.0  2013-02-13T02:00:00Z  
18415     4.60312        NaN     0.0       NaN   10.0  2013-02-12T05:00:00Z  
9709      4.60312        NaN     0.0    1007.1   10.0  2013-02-12T05:00:00Z  
1029      4.60312        NaN     0.0    1013.7   10.0  2013-02-13T04:00:00Z  

[72 rows x 15 columns]

There seems to be an outlier (likely inaccurate) data point at a wind speed of 1048.36058, so we'll have to remove it:

In [9]:
weather2 = weather2[(weather2.wind_speed < 1000)].sort_values(by=['wind_speed'], ascending=False)
weather2

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
18417    LGA  2013      2   12     2  42.98  26.06  50.94     290.0   
18428    LGA  2013      2   12    13  44.06  23.00  43.02     300.0   
1018     EWR  2013      2   12    12  44.06  26.06  48.87     270.0   
18429    LGA  2013      2   12    14  44.06  23.00  43.02     300.0   
1008     EWR  2013      2   12     2  39.92  28.04  62.21     270.0   
...      ...   ...    ...  ...   ...    ...    ...    ...       ...   
1006     EWR  2013      2   12     0  39.92  39.02  96.55     240.0   
1027     EWR  2013      2   12    21  35.96  24.08  61.67     300.0   
18415    LGA  2013      2   12     0  42.08  37.40  86.89       NaN   
9709     JFK  2013      2   12     0  42.08  37.04  82.17     280.0   
1029     EWR  2013      2   12    23  35.96  23.00  58.94     250.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
18417    23.01560   31.07106     0.0    1007.1   10.0  2013-02-12T07:00:00Z  
18428    21.86482   25.31716     0.0    1011.7   10.0  2013-02-12T18:00:00Z  
1018     21.86482   31.07106     0.0    1012.5   10.0  2013-02-12T17:00:00Z  
18429    20.71404   25.31716     0.0    1011.5   10.0  2013-02-12T19:00:00Z  
1008     20.71404   25.31716     0.0    1007.8   10.0  2013-02-12T07:00:00Z  
...           ...        ...     ...       ...    ...                   ...  
1006      6.90468        NaN     0.0    1006.9   10.0  2013-02-12T05:00:00Z  
1027      5.75390        NaN     0.0    1013.0   10.0  2013-02-13T02:00:00Z  
18415     4.60312        NaN     0.0       NaN   10.0  2013-02-12T05:00:00Z  
9709      4.60312        NaN     0.0    1007.1   10.0  2013-02-12T05:00:00Z  
1029      4.60312        NaN     0.0    1013.7   10.0  2013-02-13T04:00:00Z  

[71 rows x 15 columns]

Because there are multiple wind speed recordings for each airport on that day, we'll have to calculate the average wind speed for each airport:

In [10]:
weather3 = weather2.groupby(['origin']).mean().sort_values(by=['wind_speed'], ascending=False)
weather3

year  month   day       hour       temp       dewp      humid  \
origin                                                                    
LGA     2013.0    2.0  12.0  11.500000  41.937500  24.305000  49.937917   
JFK     2013.0    2.0  12.0  11.500000  41.247500  25.295000  53.634583   
EWR     2013.0    2.0  12.0  11.869565  40.624348  26.396522  58.029130   

          wind_dir  wind_speed  wind_gust  precip     pressure   visib  
origin                                                                  
LGA     279.130435   14.960140  24.885618     0.0  1011.113043  10.000  
JFK     279.166667   14.384750  24.584845     0.0  1011.479167   9.875  
EWR     269.130435   13.258987  25.173313     0.0  1011.628571  10.000

We can see that with an average of 14.960140, **La Guardia Airport (LGA)** had the highest average wind speed for February 12, 2013.